### ISSUE TITLE
Fallback software renderer request

### ISSUE THREAD

**Issue by tomrow** [LLK/scratch-gui#2427]
Would it be possible to add a fallback software renderer if WebGL is not set up on a user's browser? I don't think recommending that people toggle [GPU blacklisting](chrome://flags/#ignore-gpu-blacklist) in the Chrome flags page would be a good idea.

**Comment by thisandagain** [comment]
Agreed! It's a pretty significant effort though. We'd love see some contributions from the community to help support this.

**Comment by tomrow** [comment]
I wonder what license Phosphorus has, because a fork of Phosphorus player may be able to be used as a substitute for scratch-vm, and Blockly (vanilla at least) doesn't require WebGL to render blocks.

**Comment by thisandagain** [comment]
Phosphorus is a Scratch to Javascript converter, not an interpreter. While parts of the rendering engine may be able to be extricated, the two architectures are fundamentally like comparing apples and elephants.

**Comment by adroitwhiz** [comment]
I'm not sure who's getting notifications on this issue anymore, but a while ago I created [a canvas-based fork of scratch-render](https://github.com/adroitwhiz/fabricat). This doesn't implement graphic effects (I'm not sure how to even do that without significantly slowing things down), but it's functional on devices that don't support WebGL.
